R/XcmsExperiment-functions.R

```diff
@@ -878,40 +883,49 @@
     fd <- fData(chrs)
     rtc <- c("rtmin", "rtmax")
     mzc <- c("mzmin", "mzmax")
-    cpd <- chromPeakData(object)
-    pks_empty <- chromPeaks(object)[integer(), ]
-    pkd_empty <- cpd[integer(), ]
+    if (inherits(object, "XcmsExperiment"))
+        cpd <- .chromPeakData(object)
+    else cpd <- as.data.frame(chromPeakData(object))
+    message("Processing chromatographic peaks")
+    pb <- progress_bar$new(format = paste0("[:bar] :current/:",
+                                           "total (:percent) in ",
+                                           ":elapsed"),
+                           total = nrow(chrs) + 1L, clear = FALSE)
     for (i in seq_len(nrow(chrs))) {
-        pks <- chromPeaks(object, rt = fd[i, rtc], mz = fd[i, mzc],
-                          msLevel = chrs[i, 1]@msLevel, type = chromPeaks)
+        idx <- .index_chrom_peaks(
```

Comment:
instead of getting the filtered `chromPeaks` matrix we simply get the indices of the rows matching the filter. That allows to more efficiently subset also the `chromPeakData`.